File: quickstart_guides/arrays_strings/java/FindAnagrams.java
/*
 *   Title: Find anagrams
 *   Problem:
 *        Given a string s and a non-empty string p, find all the start indices
 *        of p's anagrams in s.
 *
 *       Strings consists of lowercase English letters only and the length of
 *       both strings s and p will not be larger than 20,100.
 *
 *       The order of output does not matter.
 *
 *   Execution: javac FindAnagrams.java && java FindAnagrams
 */
import java.util.*;


public class FindAnagrams {
    static List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int len = p.length();

        if (len>s.length()) {
            return res;
        }

        int[] table = new int[26];
        Arrays.fill(table, 0);
        char[] pch = p.toCharArray();

        for (char ch: pch){
            table[ch-'a']++;
        }

        char[] sch = s.toCharArray();
        int[] comp = new int[26];

        for (int i=0; i<len-1; i++){
            comp[sch[i]-'a']++;
        }

        boolean flag=true;
        for (int i=0; i<=s.length()-len; i++){
            comp[sch[i+len-1]-'a']++;
            for (int j=0; j<=25; j++){
                if(table[j]!=comp[j]){
                    flag=false;
                    break;
                }
            }
            if (flag) {
                res.add(i);
            }
            flag=true;
            comp[sch[i]-'a']--;
        }
        return res;
    }

    public static boolean check_array_equality(List<Integer> arr_1, List<Integer> arr_2) {

        if (arr_1.size() != arr_2.size()) {
            return false;
        }

        for (int i = 0; i < arr_1.size(); i++) {
            if (arr_1.get(i) != arr_2.get(i)) {
                return false;
            }
        }
        return true;
    }

    // Sample test cases
    public static void main(String[] args) {

        // Test 1:
        List<Integer> expected_output_1 = new ArrayList<Integer>();
        expected_output_1.add(0);
        expected_output_1.add(6);

        System.out.println(check_array_equality(findAnagrams("cbaebabacd", "abc"), expected_output_1));

        // Test 2:
        List<Integer> expected_output_2 = new ArrayList<Integer>();
        expected_output_2.add(0);
        expected_output_2.add(1);
        expected_output_2.add(2);

        System.out.println(check_array_equality(findAnagrams("abab", "ab"), expected_output_2));

        System.out.println("Passed all test cases");
    }
    
}

File: quickstart_guides/arrays_strings/python/find_anagrams.py
"""
Title: Find anagrams

Problem:
    Given a string s and a non-empty string p, find all the start indices of
    p's anagrams in s.

    Strings consists of lowercase English letters only and the length of both
    strings s and p will not be larger than 20,100.

    The order of output does not matter.

Execution: python find_anagrams.py
"""
from collections import Counter
from typing import List
import unittest


def find_anagrams(s: str, p: str) -> List[int]:
    pattern_counter = Counter(p)
    running_counter = Counter()
    len_p = len(p)
    result = []

    for i in range(len(s)):

        # If index  >= length of the pattern.  then decrement the count of the
        # (i - len_p)th character to remove it from the current (sliding)
        # window.
        if i >= len_p:
            running_counter[s[i - len_p]] -= 1

            if running_counter[s[i - len_p]] == 0:
                del running_counter[s[i - len_p]]

        # Default: just increment the count of the current character.
        running_counter[s[i]] += 1

        # At any time, if running_counter == pattern_counter then append the
        # result.
        if running_counter == pattern_counter:
            result.append(i - len_p + 1)

    return result


class TestCompareVersion(unittest.TestCase):
    """Unit tests for compare_version."""

    def test_1(self):
        self.assertEqual(find_anagrams("cbaebabacd", "abc"), [0, 6])

    def test_2(self):
        self.assertEqual(find_anagrams("abab", "ab"), [0, 1, 2])


if __name__ == "__main__":
    unittest.main()
